File: graphs/2/toposort/toposort.py
#Uses python3

import sys

def dfs(adj, used, order, removed, x):
    used[x] = True
    for c in adj[x]:
        if removed[c]: continue
        if used[c]:
            raise Exception(x,c)
        else:
            dfs(adj,used,order,removed,c)
    removed[x] = True
    order.append(x)
    #write your code here
    # pass


def toposort(adj):
    n = len(adj)
    used = [0] * n
    order = []
    removed = [0] * n
    adj_reverse = [[] for i in range(n)]
    for i in range(n):
        for j in range(len(adj[i])):
            adj_reverse[adj[i][j]].append(i)
    #write your code here
    for i in range(n):
        if not removed[i]:
            dfs(adj_reverse,used,order,removed,i)
    return order

if __name__ == '__main__':
    # sys.stdin = open(r'C:\Users\Admin\Desktop\crsra\graphs\2\toposort\test\01', "r")
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
    order = toposort(adj)
    for x in order:
        print(x + 1, end=' ')

File: graphs/2/toposort/toposort.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <string>
using namespace std;
bool used[(int)1e3+2], removed[(int)1e3+2];
int n,m;

bool dfs(int v, vector<vector<int> >&adj, vector<int>&order){
  used[v] = 1;
  for (int i = 0; i < adj[v].size();i++){
    int next = adj[v][i];
    if (removed[next]) continue;
    if (!used[next])
      dfs(next, adj, order);
  }
  removed[v] = 1;
  order.push_back(v);
  return 1;
}   

vector<int> toposort(vector<vector<int> > adj) {
  vector<int> order;
  vector<vector<int> > adj_reverse(n);
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < adj[i].size(); j++) {
      adj_reverse[j].push_back(i);
    }
  }
  for (int i = 0; i < n; i++){
    if (!removed[i]) {
      dfs(i, adj_reverse, order);
    }
  }
  for (int i = 0; i < order.size(); i++){
    cout << order[i] << ' ';
  }
  return order;
}

int main() {
  // size_t n, m;
  string file_path = __FILE__;
  string dir_path = file_path.substr(0, file_path.rfind("\\"));
  string test_path = dir_path + "\\test\\01";
  freopen(test_path.c_str(), "r", stdin);

  
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  for (size_t i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
  }
  vector<int> order = toposort(adj);
  for (size_t i = 0; i < order.size(); i++) {
    std::cout << order[i] + 1 << " ";
  }
}
